### PR TITLE
Feat: ensure that customers are not billed twice

### PR DESCRIPTION
**Pull request by DGajewska**
WATER-2822

* adds migrations to remove billing_charge_versions table and add transaction
* adds migrations to add transaction_type and is_summer to billing_charge_version_years table
* removes all code linked to billing charge versions
* flags licences removed during TPT review for supplementary billing

### CODE REVIEW

File: migrations/sqls/20201022143252-remove-batch-charge-versions-table-up.sql [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+      and y.financial_year_ending <= b.to_financial_year_ending
+    ) as tpt_batch
+    join (select y.billing_batch_id, y.charge_version_id, y.financial_year_ending, y.status
+      from water.billing_batch_charge_version_years as y
+      join water.billing_batches as b on y.billing_batch_id=b.billing_batch_id	
+      join water.billing_invoices as i on b.billing_batch_id=i.billing_batch_id 
+      join water.billing_invoice_licences as il on i.billing_invoice_id=il.billing_invoice_id 
+      join water.billing_transactions as t on il.billing_invoice_licence_id=t.billing_invoice_licence_id
+      where y.transaction_type is null
+        and t.is_two_part_tariff_supplementary=true
```
Comment (james-siteclick) [inline migrations/sqls/20201022143252-remove-batch-charge-versions-table-up.sql:42]: Do we need:
`and b.batch_type='supplementary'` 
here?

File: migrations/sqls/20201022143252-remove-batch-charge-versions-table-up.sql [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+      group by y.billing_batch_id, y.charge_version_id, y.financial_year_ending, y.status
+    ) as supp_batch 
+  on tpt_batch.charge_version_id=supp_batch.charge_version_id;
+
+/* Create charge version years records for TPT transactions which were not part of a TPT run 
+   ie mop up the charge version years which were missed in above query */
+insert into water.billing_batch_charge_version_years (billing_batch_id, charge_version_id, financial_year_ending, status, transaction_type, is_summer)
```
Comment (james-siteclick) [inline migrations/sqls/20201022143252-remove-batch-charge-versions-table-up.sql:49]: I don't understand why a second query is needed - again be good to chat through

**Review by james-siteclick — CHANGES_REQUESTED** [review]

**Review by james-siteclick — COMMENTED** [review]
I think I had expected to see some changes to the charge processor code but I can't spot any here

File: migrations/sqls/20201028120146-billing-invoice-licences-constraint-changes-up.sql [hunk 3]
```diff
@@ -1,6 +1,7 @@
-ALTER TABLE water.billing_invoice_licences ADD CONSTRAINT billing_invoice_licences_unique UNIQUE (billing_invoice_id,licence_id);
+ALTER TABLE water.billing_invoice_licences DROP CONSTRAINT IF EXISTS unique_batch_invoice_licence;
 ALTER TABLE water.billing_invoice_licences DROP COLUMN company_id;
 ALTER TABLE water.billing_invoice_licences DROP COLUMN contact_id;
 ALTER TABLE water.billing_invoice_licences DROP COLUMN address_id;
 ALTER TABLE water.billing_invoice_licences DROP COLUMN licence_holder_name;
 ALTER TABLE water.billing_invoice_licences DROP COLUMN licence_holder_address;
+ALTER TABLE water.billing_invoice_licences ADD CONSTRAINT billing_invoice_licences_unique UNIQUE (billing_invoice_id,licence_id);
```
Comment (james-siteclick) [inline migrations/sqls/20201028120146-billing-invoice-licences-constraint-changes-up.sql:7]: Is this related to the recent work on ensuring only 1 invoiceLicence per licence per invoice?

File: src/lib/connectors/repos/billing-batch-charge-version-years.js [hunk 4]
```diff
@@ -2,7 +2,34 @@
...
 const raw = require('./lib/raw');
+const helpers = require('./lib/helpers');
 const queries = require('./queries/billing-batch-charge-version-years');
+const { TRANSACTION_TYPE } = require('../../models/charge-version-year');
+
+/**
+ * Finds a charge version year with related model data by ID
+ * @param {String} id
+ */
+const findOne = async id => {
```
Comment (james-siteclick) [inline src/lib/connectors/repos/billing-batch-charge-version-years.js:13]: I think there is a `fineOne` helper which allows this to be simplified

File: src/lib/connectors/repos/billing-batch-charge-version-years.js [hunk 5]
```diff
@@ -53,11 +80,19 @@ const findByBatchId = async billingBatchId => {
  * @param {String} billingBatchId
  * @return {Promise<Array>}
  */
-const findTwoPartTariffByBatchId = billingBatchId =>
-  raw.multiRow(queries.findTwoPartTariffByBatchId, { billingBatchId });
+const findTwoPartTariffByBatchId = async billingBatchId => {
```
Comment (james-siteclick) [inline src/lib/connectors/repos/billing-batch-charge-version-years.js:83]: I think the `findMany` helper could be used here too

File: src/modules/billing/services/charge-processor-service/index.js [hunk 6]
```diff
@@ -31,24 +31,26 @@ const createInvoiceLicence = (chargeVersion) => {
...
-  if (!chargeVersion) {
-    throw new NotFoundError(`Charge version ${chargeVersionId} not found`);
-  }
+const processChargeVersionYear = async (chargeVersionYear) => {
+  validateData(chargeVersionYear);
 
+  const { chargeVersion, financialYear } = chargeVersionYear;
   // Load all billing volumes relating to the charge elements and financial year
   // For supplementary these could have been generated by a previous TPT batch, not the current batch
   // There could also be multiple billing volumes per element - for summer and winter/all year TPT batches
```
Comment (james-siteclick) [inline src/modules/billing/services/charge-processor-service/index.js:56]: Could we skip loading the billingVolumes here if the `chargeVersionYear.transactionType` is `annual`?

File: src/modules/billing/services/charge-version-year.js [hunk 7]
```diff
@@ -1,16 +1,25 @@
...
+const { CHARGE_VERSION_YEAR_STATUS, TRANSACTION_TYPE } = require('../../../lib/models/charge-version-year.js');
+
+/**
+ * Gets charge version year for given id
+ * @param {String} id
+ * @return {Promise}
+ */
+const getChargeVersionYearById = async id => {
+  const row = await repos.billingBatchChargeVersionYears.findOne(id);
+  return row ? mappers.chargeVersionYear.dbToModel(row) : null;
```
Comment (james-siteclick) [inline src/modules/billing/services/charge-version-year.js:21]: I think there is a `findOne` service helper that could maybe be used for this?

**Review by james-siteclick — CHANGES_REQUESTED** [review]

File: migrations/sqls/20201028120146-billing-invoice-licences-constraint-changes-up.sql (hunk 3 above)
Comment (DGajewska) [reply migrations/sqls/20201028120146-billing-invoice-licences-constraint-changes-up.sql:7]: This was an issue when I was moving up and down through migrations when billing data was present in my local database. Is it incorrect?

Comment (james-siteclick) [reply migrations/sqls/20201028120146-billing-invoice-licences-constraint-changes-up.sql:7]: I have tested locally and they seem to work fine for me too

**Review by james-siteclick — APPROVED** [review]

**Pull request merged**